### PR TITLE
[Bug] Improve SSO user provision flow

### PR DESCRIPTION
**Pull request by vincentsalucci**
## Code Changes
**Sso/AccountController**
- Adjusted `guid` to not be nullable - removed downstream checks. 
- Added more robust `orgUser` checking. Will look by `user.Id` if available or will try to search by `email` via new stored procedure.
- Clarified the `existingUser` flow: adjusted error messages and added comments for all situations
- Removed user count check and isolated seat count check before any user creation
- Updated `orgUser` creation by adding conditional for an update if the `orgUser` already exists (adds `user.id` to `orgUser`)
- Created helper method for creation of SSO object

**IOrganizationUserRepository**
- Added `GetByOrganizationEmailAsync` method reference

**OrganizationUserRepository**
- Added `GetByOrganizationEmailAsync` method that will call a new stored procedure and return an `orgUser` within the given `orgId` and has a matching `email`

**OrganizationUser_ReadByOrganizationIdEmail**
- Created stored procedure to return a user within the given organization and has a matching email

**2020_12_04_00_OrgUserReadByOrgEmail**
- Created migrator script

**SharedResources**
- Updated `UserAlreadyInvited` string.  Added some follow-up steps if the user runs into this error.
- Added `UserAlreadyExistsInviteProcess`. This occurs when the user exists but there is **no** `orgUser` found.

### CODE REVIEW

File: bitwarden_license/src/Sso/Controllers/AccountController.cs
Comment (cscharf) [inline bitwarden_license/src/Sso/Controllers/AccountController.cs:412]: I think you may want to split this up into 2 separate DB calls; the list of org users could be in the 1k's, really you need to get a single by User Id, if currentUser is not null, then if that result is null OR current user is null, fetch by Email (separate call).

The logic I see here looks like it will apply an OR, what if there is an OrgUser with the UserID *and* a different OrgUser with that email address? The logic could not distinguish them nor give preference to the UserId match first.

File: bitwarden_license/src/Sso/Controllers/AccountController.cs
Comment (cscharf) [inline bitwarden_license/src/Sso/Controllers/AccountController.cs:542]: Why not have this method literally create the record and persist it (making it async)? That will extend the re-use even more so.

**Review by cscharf — CHANGES_REQUESTED** [review]
Just a couple of things, but really, really close I think. We can discuss tomorrow if necessary.

File: bitwarden_license/src/Sso/Controllers/AccountController.cs
Comment (vincentsalucci) [reply bitwarden_license/src/Sso/Controllers/AccountController.cs]: Updated! Split into two separate db calls and prioritized `existingUser.Id` search result.

File: bitwarden_license/src/Sso/Controllers/AccountController.cs
Comment (cscharf) [inline bitwarden_license/src/Sso/Controllers/AccountController.cs:421]: Ah, okay, this is where we discussed creating a new stored procedure to find an org user by OrgId + Email, vs. reading all the org users and filtering further by email.... not sure I'm going to necessarily be a stickler on that, while performance will take a bit of a hit for larger organization.... most of those organizations aren't going to be going through this "flow" very often, so it won't be high-frequency enough to spend too much time factoring for performance up-front.

**Review by cscharf — DISMISSED** [review]
Awesome! Please don't merge it yet!

**Comment by cscharf** [comment]
Good to merge this when you're ready.

File: bitwarden_license/src/Sso/Controllers/AccountController.cs
Comment (vincentsalucci) [reply bitwarden_license/src/Sso/Controllers/AccountController.cs]: Updated - helper method now persists the `sso user` creation

File: bitwarden_license/src/Sso/Controllers/AccountController.cs
Comment (vincentsalucci) [reply bitwarden_license/src/Sso/Controllers/AccountController.cs]: Sproc created. Conditional updated.

File: bitwarden_license/src/Sso/Controllers/AccountController.cs
Comment (cscharf) [inline bitwarden_license/src/Sso/Controllers/AccountController.cs:419]: ooh, sexy syntax! 🌶️

**Review by cscharf — APPROVED** [review]

File: util/Migrator/DbScripts/2020_12_04_00_OrgUserReadByOrgEmail.sql
Comment (cscharf) [inline util/Migrator/DbScripts/2020_12_04_00_OrgUserReadByOrgEmail.sql:1]: @vincentsalucci, This check and `DROP` applies to `OrganizationUser_ReadByOrganizationEmail`; however the `CREATE` below refers to `OrganizationUser_ReadByOrganizationIdEmail` which is different.

Just noticed this as it's breaking my deployment so we'll want to get this corrected in a new PR.

**Pull request merged**
